### ISSUE TITLE
Suggestions for modelling

### ISSUE THREAD

**Issue by ishaanmht** [google-deepmind/PGMax#7]
I am working on a resource allocation problem using PGMax. In this problem, I have m agents and  n tasks, where 
n>m. I need to assign each agent to a task, and ideally, each agent should be matched with the closest available task on a 2D grid to minimize the distance. My constraints are that no two agents can be assigned the same task, and agents have the option to choose no task at all, which incurs a penalty.

I have implemented a model using PGMax, but I am unsure if my implementation is correct. Specifically, I’ve noticed that the output values of the variables remain unchanged even when I modify the cost structure, such as the distances between agents and tasks. Additionally, there are instances where multiple agents end up being assigned to the same task, which is not desired.

Any guidance or advice would be greatly appreciated. I have attached my code for further review.

`
import jax
import jax.numpy as jnp
import matplotlib.pyplot as plt
import numpy as np

# Load PGMax
from pgmax import fgraph, fgroup, infer, vgroup

# Parameters
m = 5  # Number of agents
n = 10 # Number of tasks (n > m)
penalty =  -10000  # Large penalty for selecting the same task
idle_penalty = -100  # Penalty for staying idle

# Simulate 2D positions for agents and tasks
np.random.seed(42)
agent_positions = np.random.rand(m, 2) * 10  # m agents in a 10x10 space
task_positions = np.random.rand(n, 2) * 10  # n tasks in a 10x10 space

# Calculate the cost matrix (m x n) for agents selecting tasks
cost_matrix = np.zeros((m, n))
for i in range(m):
    for j in range(n):
        cost_matrix[i, j] = np.linalg.norm(agent_positions[i] - task_positions[j])

# Add an extra state for each agent to stay idle (with an idle penalty)
cost_matrix = np.hstack([cost_matrix, np.full((m, 1), idle_penalty)])

# Define the variables
variables = vgroup.NDVarArray(num_states=n + 1, shape=(m,))

# Create the Factor Graph
fg = fgraph.FactorGraph(variable_groups=[variables])

# Create the pairwise factor group
variables_for_factors = []
factor_matrices = []

# Prepare the 3D matrix
num_pairs = m * (m - 1) // 2
factor_matrix_3d = np.zeros((num_pairs, n + 1, n + 1))

pair_index = 0
for i in range(m):
    for j in range(i + 1, m):
        # Create a factor matrix for the pair (i, j)
        factor_matrix = np.zeros((n + 1, n + 1))
        for state_i in range(n + 1):
            for state_j in range(n + 1):
                if state_i == state_j and state_i != n:  # Same task (not idle)
                    factor_matrix[state_i, state_j] = penalty
                else:
                    factor_matrix[state_i, state_j] = (cost_matrix[i, state_i] + cost_matrix[j, state_j])

        # Store the factor matrix in the 3D matrix
        factor_matrix_3d[pair_index] = factor_matrix

        # Record variables for this factor
        variables_for_factors.append([variables[i], variables[j]])

        pair_index += 1

print(factor_matrix_3d)
# Create the Pairwise Factor Group
factor_group = fgroup.PairwiseFactorGroup(variables_for_factors=variables_for_factors,
    log_potential_matrix=factor_matrix_3d,
)

# Add the pairwise factors to the factor graph
fg.add_factors(factor_group)

# Update the evidence
rng = jax.random.PRNGKey(2)
evidence_updates={variables: jax.random.gumbel(rng, shape=(m, n+1))}

# Perform Loopy Belief Propagation
bp = infer.build_inferer(fg.bp_state, backend="bp")

# Run MAP inference
inferer_arrays = bp.init(evidence_updates=evidence_updates)
inferer_arrays, msgs_deltas = bp.run_with_diffs(inferer_arrays, num_iters=2000, temperature=0.0)

# Compute the beliefs
beliefs = bp.get_beliefs(inferer_arrays)

print(beliefs)
# Extract the marginals (beliefs)
marginals = infer.decode_map_states(beliefs)
print(marginals)
# Print out the marginals
for var_id, marginal in enumerate(marginals):

    print(f"Agent {var_id + 1} Beliefs: {marginal}")

# Plot agent and task positions for visualization
plt.figure(figsize=(8, 8))
plt.scatter(agent_positions[:, 0], agent_positions[:, 1], c='blue', label='Agents', s=100)
plt.scatter(task_positions[:, 0], task_positions[:, 1], c='red', label='Tasks', s=100)
for i in range(m):
    plt.text(agent_positions[i, 0], agent_positions[i, 1], f'Agent {i+1}', fontsize=12)
for j in range(n):
    plt.text(task_positions[j, 0], task_positions[j, 1], f'Task {j+1}', fontsize=12)
plt.legend()
plt.title("Agent and Task Positions")
plt.show()`
